bitcoin/src/blockdata/script/mod.rs

```diff
@@ -680,6 +736,148 @@ pub(super) fn bytes_to_asm_fmt(script: &[u8], f: &mut dyn fmt::Write) -> fmt::Re
+        ParseAsmError { position, kind }
+    }
+
+    let mut buf = Vec::with_capacity(65);
+    let mut builder = Builder::new();
+    let mut words = iter_words(asm);
+    while let Some((pos, mut word)) = words.next() {
+        // We have this special case in our formatter.
+        if word == "OP_0" {
+            builder = builder.push_opcode(OP_PUSHBYTES_0);
+            continue;
+        }
+
+        if let Ok(op) = Opcode::from_str(word) {
+            // check for push opcodes
+            if op.code <= OP_PUSHDATA4.code {
+                let (next, push) = words.next().ok_or(err(pos, AsmParseErrorKind::UnexpectedEOF))?;
+                if !try_parse_raw_hex(push, &mut buf) {
+                    return Err(err(next, AsmParseErrorKind::InvalidHex));
+                }
+
+                // NB our API doesn't actually allow us to make byte pushes with
+                // non-minimal length prefix, so we can only check and error if
+                // the user wants a non-minimal push
+                let expected_push_op = match buf.len() {
+                    n if n < opcodes::all::OP_PUSHDATA1.code as usize => {
+                        Opcode::from(n as u8)
+                    }
+                    n if n < 0x100 => {
+                        opcodes::all::OP_PUSHDATA1
+                    }
+                    n if n < 0x10000 => {
+                        opcodes::all::OP_PUSHDATA2
+                    }
+                    n if n < 0x100000000 => {
+                        opcodes::all::OP_PUSHDATA4
+                    }
+                    _ => return Err(err(next, AsmParseErrorKind::PushExceedsMaxSize)),
+                };
+                if op != expected_push_op {
+                    return Err(err(pos, AsmParseErrorKind::NonMinimalBytePush));
+                }
+
+                let push = <&PushBytes>::try_from(&buf[..])
+                    .map_err(|_| err(next, AsmParseErrorKind::PushExceedsMaxSize))?;
+                builder = builder.push_slice(push);
+            } else {
+                builder = builder.push_opcode(op);
+            }
+            continue;
+        }
+
+        // Not an opcode, try to interpret as number or push.
+
+        if word.starts_with('<') && word.ends_with('>') {
+            word = &word[1..word.len()-1];
+        }
+
+        // Try a number.
+        if let Ok(i) = i64::from_str(&word) {
+            builder = builder.push_int(i);
+            continue;
+        }
+
+        // Finally, try hex in various forms.
+        if word.starts_with("0x") {
+            word = &word[2..];
+        }
+
+        if try_parse_raw_hex(word, &mut buf) {
+            let push = <&PushBytes>::try_from(&buf[..])
+                .map_err(|_| err(pos, AsmParseErrorKind::PushExceedsMaxSize))?;
+            builder = builder.push_slice(push);
+        } else {
+            return Err(err(pos, AsmParseErrorKind::UnknownInstruction));
+        }
+    }
+
+    Ok(builder.into_script())
+}
```

Comment:
This is interesting but it makes our existing "asm" output first class. We should maybe think about whether we want to change it (bearing in mind that it's currently heavily used, at least in user display, in e.g. Esplora).

* Do we want to encode decimal numbers at all?
* Do we want to drop the `OP_` prefix from opcodes?
* Should the `OP_PUSHNUM_x` names be shortened to just `OP_x`? How about the `OP_PUSHBYTES_n`?
* Should we handle `OP_TRUE` as an alias for `OP_1`? `OP_FALSE` for `OP_0`? `OP_PUSHBYTES_0` for `OP_0`? etc
* Should we allow the old `NOPx` names for CSV and CLTV?

We don't need to answer these here ofc but there will be a long debate on the actual PR that introduces this.